keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -1188,239 +937,309 @@ drivers are expected to be written for Kubernetes.
 down and then gets removed.
 
 All list types are atomic because that makes tracking the owner for
 server-side-apply (SSA) simpler. Patching individual list elements is not
-needed and there is a single owner (kubelet).
+needed and there is a single owner.
 
 ```go
-// ResourceSlice provides information about available
-// resources on individual nodes.
+// One of more slices represent a pool of devices managed by a given driver. For
+// node-local devices, each node is a pool. How many slices the driver uses
+// to publish that pool is driver-specific. Within each pool the device name must
+// be unique.
+//
+// In which slice a device gets published may change over time. The unique identifier
+// for a node-local device is the tuple `<driver name>/<node name>/<device name>`. Each
+// of these names is a DNS label or domain, so it is okay to concatenate them
+// like this in a string with a slash as separator.
+//
+// Whenever a driver needs to update a pool, it bumps the pool generation number
+// and updates the slices with that new number and new device definitions. A consumer
+// must only use device definitions from the slices with the highest generation number
+// and ignore the others.
+//
+// If necessary, a consumer can check the number of total devices in a pool (included
+// in each slice) to determine whether its view of a pool is complete.
+//
+// For devices that are not local to a node, the node name is not set. Instead,
+// a pool name is chosen by the driver if necessary. It can be left blank if the
+// device names are unique in the entire cluster. The unique identifier then
+// has the same format as before with `<driver name>[/<pool name>]/<device name>`.
 type ResourceSlice struct {
     metav1.TypeMeta
     // Standard object metadata
+    // +optional
     metav1.ObjectMeta
 
-    // NodeName identifies the node which provides the resources
-    // if they are local to a node.
+    Spec ResourceSliceSpec
+
+    // A status might get added later.
+}
+
+type ResourceSliceSpec struct {
+    // NodeName identifies the node which provides the devices. All devices
+    // are local to that node and part of a per-node pool.
     //
-    // A field selector can be used to list only ResourceSlice
-    // objects with a certain node name.
+    // This field can be used to limit access from nodes to slices with
+    // the same node name. It also indicates to autoscalers that adding
+    // new nodes of the same type as some old node might also make new
+    // devices available.
+    //
+    // NodeName and PoolName are mutually exclusive.
+    //
+    // +optional
     NodeName string
 
+    // PoolName can be used to describe pools of devices which are not local
+    // to a node. It can be left blank if the device names are unique in the cluster.
+    //
+    // It must not be longer than 253 and must consist of one or more DNS sub-domains
+    // separated by slashes.
+    //
+    // +optional
+    PoolName string
```

Comment:
Do we really need two names? Can't `PoolName` just be the `NodeName` in cases where all devices are node local?

Reply:
We can group "PoolName" and "NodeSelector" under "Pool", a struct containing "Name" and "NodeSelector".

The "NodeName" and "Pool" are properly mutually exclusive.

Reply:
I guess the confusing thing to me is that whether we are node-local or not, we still talk about having a "pool" of devices. Which (putting on my hat as a new-comer to this) would imply to me that I should have a `PoolName` in both cases. 

I don't have a suggestion off the top of my head to avoid this (or an alternate name for this), but it's worth noting how confusing this is.

Reply:
Could we require both `PoolName` and `NodeSelector` and in the single node case, set `NodeSelector` to the one and only node where the devices are available?

Reply:
If we do that, then it's unclear whether such a `ResourcePoolSlice` is about node-local resources.

A driver might manage a fabric-attached device which happens to be available only on one node. We cannot infer from `NodeSelector` to "local to this node", which is relevant for authorizing kubelet and node-local drivers.

We would need a boolean to make that unambiguous:
- `PoolName`: name of the pool, optional for network-attached devices
- `NodeLocal`: if true, the devices are local to a node and the `PoolName` has to be the node name; has to be available for field filters
- `NodeSelector`: optional, mutually exclusive with `NodeLocal`.

Then when kubelet does its deletecollection, it has to use `poolName=<node name>,nodeLocal=true`.

Reply:
OK. this is worse.

The main thing I don't like about the way things are currently proposed is that the canonical name can either be:
1. `<driver name>/<node name>/<device name>`; OR
1. `<driver name>/<pool name>/<device name>`; OR
1. `<driver name>/<device name>`

There must be some way to ensure the canonical name is always built in the same way AND ensure that when devices are node-local that the kubelet and node-local drivers are properly authorized.

Reply:
Yes, but which way? :cry: 

I now tend to describe the device ID as `<driver name>[/<pool name>]/<device name>`, with `<pool name>` == `<node name>` as special case for local devices. Not sure whether that makes it easier to digest.

How about this:
```
    // For node-local devices, this is the name of their node. It also serves as the pool name.
    //
    // NodeName and NonLocal are mutually exclusive.
    //
    // +optional
    NodeName string

    // NonLocal must be set when the devices are not local to a node.
    //
    // NodeName and NonLocal are mutually exclusive.
    //
    // +optional
    NonLocal *NonLocal

...

type NonLocal struct {
   // The pool name if device names are not globally unique.
   // +optional
   PoolName string

   // Where devices can be used. If nil, they can be used in the entire cluster.
   // +optional
   NodeSelector *v1.NodeSelector
}
```

Reply:
What about:
1. `PoolName` is required
2. `NodeName` and `NodeSelector` are mutually exclusive. 
3. If you're node-local you set `NodeName`
4. If you're network attached, you set `NodeSelector`
```
This then give the canonical name of `<driver name>/<pool name>/<device name>` in all cases and provides some symmetry with `NodeName` vs. `NodeSelector` in the name of the field you have to set for node-local vs. network-attached devices.

Reply:
Do we require that "PoolName" == "NodeName" if both are set or do we only make it a suggestion?

I'm fine either way.

Reply:
Does `PoolName` really have to be required? It might not be needed when device names are globally unique.

Reply:
I'm fine with requiring it, driver authors can always provide "global" if they cannot leave it empty - I just want to be sure.

Reply:
I would say PoolName == NodeName is a suggestion (but it doesn’t have to be so long as it is globally unique) and yes it should be required so we don’t have to over explain the canonical naming convention.